docs: add README and fix charmcraft.yaml links (#11)
charmed-hpc/slurm-charms: README.md

Reviewer: We should style this README similar to the one that we have for the [docs](https://github.com/charmed-hpc/docs) repo. It's based off a template/gist I put together [here](https://gist.github.com/NucciTheBoss/91be6082e10a3563af92f467796be86e), which is in turn based off of other READMEs that I have seen from popular projects.

My primary reasoning for wanting to use this template is that:

1. I think it has a nice flow without going into too much detail, and it doesn't overload new people to the project.
2. It feels more inviting imho. Maybe that's just because we unabashedly use emojis, but I want folks to feel excited about trying out the Slurm charms for their use case.

We're kind of already there with this README; we just need to massage it into the following sections:

1. __Getting started__ &mdash; short how-to for how to deploy the Slurm charms.
2. __What's next__ &mdash; point folks to where they can get more information. E.g. URL to the documentation website, where they can report issues, how to get in contact with other Slurm charm users.
3. __Development__ &mdash; quick info about how to start developing. E.g. ensure the all tox envs pass before opening a PR and link to CONTRIBUTING.md for further information.
4. __Project and community__ &mdash; same as current section but different header.
5. __License__ &mdash; license info for both the Slurm charms and Slurm itself.

> <div align="center">

Reviewer: I used to like centering the description content, but I think it looks better justified to the left. Let's remove this div.

> # slurm charms

Reviewer: Slurm should be capitalized here:

```suggestion
# Slurm charms
```

> Manage HPC clusters using [Juju](https://juju.is) and the [SLURM](https://slurm.schedmd.com/overview.html) workload manager.

Reviewer: You could replace this with the tagline we have as the repository description:

```suggestion
[Juju](https://juju.is) charms for automating the Day 0 to Day 2 operations of the [Slurm workload manager](https://slurm.schedmd.com/overview.html) ⚖️🐧
```

[![CI](https://github.com/charmed-hpc/slurm-charms/actions/workflows/ci.yaml/badge.svg)](https://github.com/charmed-hpc/slurm-charms/actions/workflows/ci.yaml/badge.svg)
[![Release](https://github.com/charmed-hpc/slurm-charms/actions/workflows/release.yaml/badge.svg)](https://github.com/charmed-hpc/slurm-charms/actions/workflows/release.yaml/badge.svg)

> [![Matrix](https://img.shields.io/matrix/ubuntu-hpc%3Amatrix.org?logo=matrix&label=ubuntu-hpc)](https://matrix.to/#/#ubuntu-hpc:matrix.org)

Reviewer: We can use the ubuntu.com invite link. The reason we have the matrix.org link is because Ubuntu HPC existed on Matrix before the Ubuntu home server did :sweat_smile: 

```suggestion
[![Matrix](https://img.shields.io/matrix/ubuntu-hpc%3Amatrix.org?logo=matrix&label=ubuntu-hpc)](https://matrix.to/#/#hpc:ubuntu.com)
```

</div>

> The `slurm-charms` repository provides many of the SLURM daemons that are required to deploy, configure and manage a SLURM-based HPC cluster:

Reviewer: SLURM is the old spelling for Slurm, back when Slurm was marketed as a "Simple Linux Utilility Resource Manager." Now the preferred spelling is `Slurm` and it is marketed as the "Slurm workload manager".

I even have a Slurm t-shirt from Schedmd where the slogan is "Slurm - it's not an acronym"

- [`slurmctld`](./charms/slurmctld/): the central management daemon for SLURM.
- [`slurmd`](./charms/slurmd): the compute node daemon for SLURM.
- [`slurmdbd`](./charms/slurmdbd): the database daemon for SLURM.
- [`slurmrestd`](./charms/slurmrestd/): the REST API interface to SLURM.

> ## Usage

Reviewer: ```suggestion
## ✨ Getting started
```

This should be run using Juju 3.x or greater.

> #### Deploy a minimal Charmed SLURM cluster

Reviewer: I don't think we need this header here. I feel like it implies that in the future we're going to add more instructions when really we're going to add them to the documentation website since we can go more in-depth there.

```suggestion
```

```shell
$ juju deploy slurmctld --channel edge
$ juju deploy slurmd --channel edge
$ juju deploy slurmdbd --channel edge
$ juju deploy slurmrestd --channel edge

> $ juju deploy mysql --channel 8.0/edge

Reviewer: We should switch this to the stable branch. See https://github.com/charmed-hpc/slurm-bundles/issues/40:

```suggestion
$ juju deploy mysql --channel 8.0/stable
```

$ juju deploy mysql-router slurmdbd-mysql-router --channel dpe/edge

$ juju integrate slurmctld:slurmd slurmd:slurmctld
$ juju integrate slurmctld:slurmdbd slurmdbd:slurmctld
$ juju integrate slurmctld:slurmrestd slurmrestd:slurmctld
$ juju integrate slurmdbd-mysql-router:backend-database mysql:database
$ juju integrate slurmdbd:database slurmdbd-mysql-router:database
```

> ## Project & Community

Reviewer: ```suggestion
## 🤝 Project and community
```

> `slurm-charms` is a project of the [Ubuntu HPC](https://discourse.ubuntu.com/t/high-performance-computing-team/35988)
> community. It is an open source project that is welcome to community involvement, contributions, suggestions, fixes, and
> constructive feedback. Interested in being involved with the development of the project? Check out these links below:
> 
> * [Join our online chat](https://matrix.to/#/#ubuntu-hpc:matrix.org)
> * [Contributing guidelines](./CONTRIBUTING.md)
> * [Code of conduct](https://ubuntu.com/community/ethos/code-of-conduct)
> * [File a bug report](https://github.com/charmed-hpc/slurm-charms/issues)
> * [Juju SDK docs](https://juju.is/docs/sdk)

Reviewer: How do you feel about using this boilerplate from the doc site landing page for this section?

```suggestion
The Slurm charms are a project of the [Ubuntu High-Performance Computing community](https://ubuntu.com/community/governance/teams/hpc). Interested in contributing bug fixes, patches, documentation, or feedback? Want to join the Ubuntu HPC community? You’ve come to the right place 🤩

Here’s some links to help you get started with joining the community:

* [Ubuntu Code of Conduct](https://ubuntu.com/community/ethos/code-of-conduct)
* [Contributing guidelines](./CONTRIBUTING.md)
* [Join the conversation on Matrix](https://matrix.to/#/#hpc:ubuntu.com)
* [Get the latest news on Discourse](https://discourse.ubuntu.com/c/hpc/151)
* [Ask and answer questions on GitHub](https://github.com/orgs/charmed-hpc/discussions/categories/q-a)
```

Big thing is that I think the nitty gritty development information should be stored within CONTRIBUTING.md. That's where users who want to contribute fixes and features should go to learn more about the tech we work with such as the Juju SDK and charmcraft. The Juju SDK isn't that interesting to someone who just wants to know how to deploy the Slurm charms, or ask a question on GitHub.

> ## License

Reviewer: ```suggestion
## 📋 License
```

> This project is free software, distributed under the Apache Software License, version 2.0. See the [LICENSE](./LICENSE) file for more information.

Reviewer: We should adapt this blurb from the Slurm snap repository since the charms are under a different license than Slurm itself:

```suggestion
The Slurm charms are free software, distributed under the Apache Software License, version 2.0. See the [Apache-2.0 LICENSE](./LICENSE) file for further details. The Slurm workload manager is licensed under the GNU General Public License, version 2, or any later version. See Slurm's [legal notice](https://slurm.schedmd.com/disclaimer.html) for further licensing information about Slurm.
```

```shell
juju deploy slurmctld --channel edge
juju deploy slurmd --channel edge
juju deploy slurmdbd --channel edge
juju deploy slurmrestd --channel edge
juju deploy mysql --channel 8.0/stable
juju deploy mysql-router slurmdbd-mysql-router --channel dpe/edge

juju integrate slurmctld:slurmd slurmd:slurmctld
juju integrate slurmctld:slurmdbd slurmdbd:slurmctld
juju integrate slurmctld:slurmrestd slurmrestd:slurmctld
juju integrate slurmdbd-mysql-router:backend-database mysql:database
juju integrate slurmdbd:database slurmdbd-mysql-router:database
```

## 🤔 What's next?

If you want to learn more about all the things you can do with the Slurm charms, here are some resources for you to explore:

* [Documentation](https://canonical-charmed-hpc.readthedocs-hosted.com/en/latest)
* [Open an issue](https://github.com/charmed-hpc/slurm-charms/issues/new?title=ISSUE+TITLE&body=*Please+describe+your+issue*)
* [Ask a question on Github](https://github.com/orgs/charmed-hpc/discussions/categories/q-a)

## 🛠️ Development

The project uses [tox](tox.wiki) as its command runner, which provides some useful commands that

> will definitely help while hacking on the charms:

Reviewer: [praise]: Nice :ok_hand:

```shell
tox run -e fmt # Apply formatting standards to code.
tox run -e lint # Check code against coding style standards.
tox run -e type # Type checking.
tox run -e unit # Run unit tests.
```

We also have some integration tests in place, but be aware that it requires a fairly good amount
of computer resources to run them. We usually test with at least 4 cores and 16 GB of RAM, but feel

> free to experiment!

Reviewer: [praise]: Nice wordsmithing here. I like the encouragement to experiment :star_struck:

# Slurm charms

[Juju](https://juju.is) charms for automating the Day 0 to Day 2 operations of the [Slurm workload manager](https://slurm.schedmd.com/overview.html) ⚖️🐧

[![CI](https://github.com/charmed-hpc/slurm-charms/actions/workflows/ci.yaml/badge.svg)](https://github.com/charmed-hpc/slurm-charms/actions/workflows/ci.yaml/badge.svg)
[![Release](https://github.com/charmed-hpc/slurm-charms/actions/workflows/release.yaml/badge.svg)](https://github.com/charmed-hpc/slurm-charms/actions/workflows/release.yaml/badge.svg)
[![Matrix](https://img.shields.io/matrix/ubuntu-hpc%3Amatrix.org?logo=matrix&label=ubuntu-hpc)](https://matrix.to/#/#hpc:ubuntu.com)

The `slurm-charms` repository provides many of the Slurm daemons that are required to deploy, configure and manage a Slurm-based HPC cluster:

- [`slurmctld`](./charms/slurmctld/): the central management daemon for Slurm.
- [`slurmd`](./charms/slurmd): the compute node daemon for Slurm.
- [`slurmdbd`](./charms/slurmdbd): the database daemon for Slurm.
- [`slurmrestd`](./charms/slurmrestd/): the REST API interface to Slurm.

> ## ✨ Getting started

Author: Yeah, that should definitely be our target, but we can keep iterating on it.

> ## ✨ Getting started

Reviewer: Great! We can save this as part of docs "sprint". _I think we're still doing that_

This should be run using Juju 3.x or greater.

```shell

> $ juju deploy slurmctld --channel edge
> $ juju deploy slurmd --channel edge
> $ juju deploy slurmdbd --channel edge
> $ juju deploy slurmrestd --channel edge
> $ juju deploy mysql --channel 8.0/stable
> $ juju deploy mysql-router slurmdbd-mysql-router --channel dpe/edge
> 
> $ juju integrate slurmctld:slurmd slurmd:slurmctld
> $ juju integrate slurmctld:slurmdbd slurmdbd:slurmctld
> $ juju integrate slurmctld:slurmrestd slurmrestd:slurmctld
> $ juju integrate slurmdbd-mysql-router:backend-database mysql:database
> $ juju integrate slurmdbd:database slurmdbd-mysql-router:database

Reviewer: Since we don't have the shell prompt for the `tox` commands, should we remove them here as well?

```suggestion
juju deploy slurmctld --channel edge
juju deploy slurmd --channel edge
juju deploy slurmdbd --channel edge
juju deploy slurmrestd --channel edge
juju deploy mysql --channel 8.0/stable
juju deploy mysql-router slurmdbd-mysql-router --channel dpe/edge

juju integrate slurmctld:slurmd slurmd:slurmctld
juju integrate slurmctld:slurmdbd slurmdbd:slurmctld
juju integrate slurmctld:slurmrestd slurmrestd:slurmctld
juju integrate slurmdbd-mysql-router:backend-database mysql:database
juju integrate slurmdbd:database slurmdbd-mysql-router:database
```

Ik that `markdownlint` - which we're not using - will complain if you have the shell prompt `$` but don't have any example output shown.

```

## 🤔 What's next?

> If you want to know more about the project, there are several links available for you to explore:

Author: I don't want to follow a question (the section title) with another question, so I'll modify it a bit.

* [Documentation](https://canonical-charmed-hpc.readthedocs-hosted.com/en/latest)
* [Open an issue](https://github.com/charmed-hpc/slurm-charms/issues/new?title=ISSUE+TITLE&body=*Please+describe+your+issue*)
* [Ask a question on Github](https://github.com/orgs/charmed-hpc/discussions/categories/q-a)

## 🛠️ Development

> The project uses [tox](tox.wiki) as its command runner. It's very common to run the following four
> commands to check that the codebase is in a good state.

Reviewer: ```suggestion
This repository uses [tox](https://tox.wiki) as its command runner. Any changes you make to the Slurm charms should pass the following tests:
```

Any changes should always pass all tests :sweat_smile:

Author: This edit is already part of the contributing guide, and I think the current description is better for people experimenting with the repo.

Author: I slightly edited this.

Reviewer: I like the new wording. Nice and inviting.

```shell

> tox run -e fmt # Apply correct formatting to code.

Reviewer: I'm not sure if "correct formatting" is the right phrasing here? Wdyt about "formatting standards" instead?

```suggestion
tox run -e fmt # Apply formatting standards to code.
```

Also, nit, but I really want the comments to line up together for some reason :rofl:

Author: I'll also change tox's job name because this is just copied from there.

tox run -e lint # Check code against coding style standards.
tox run -e type # Type checking.
tox run -e unit # Run unit tests.
```

You can also run integration tests if you desire, but be aware that it requires a significant amount

> of computer resources to run them.

Reviewer: Tbh I think we can say something along the lines of _your machine should have at least 4 cores and 16 GB of RAM to run the integration tests_. That's what we have at our disposal with the [GitHub CI runner](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) we're using.

This might change later down the line, but we should measure expectations with concrete numbers.

Author: I don't want to give specific numbers because it discourages experimentation.

Author: Added some numbers, but also reworded it a bit to encourage experimentation with other setups.
